File: python/advanced-graphs/network-delay-time.py
import heapq
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        adj = defaultdict(lambda: [])
        for u, v, w in times:
            adj[u].append((w, v))

        heap = []
        distance = {}
        for i in range(1, n + 1):
            distance[i] = float('inf')

        distance[k] = 0
        heapq.heappush(heap, (0, k))
        visited = set()

        while heap:
            dist, node = heapq.heappop(heap)

            if node in visited:
                continue

            distance[node] = dist
            visited.add(node)

            for weight, neighbor in adj[node]:
                heapq.heappush(heap, (dist + weight, neighbor))

        time = float('-inf')
        for node in distance:
            time = max(distance[node], time)

        if time == float('inf'):
            return -1

        return time

File: c++/advanced-graphs/network-delay-time.cc
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        std::unordered_map<int, std::vector<std::pair<int, int>>> adj;
        for (std::vector<int> time : times) {
            adj[time[0]].push_back(std::make_pair(-time[2], time[1]));
        }

        std::priority_queue<std::pair<int, int>> pq;
        pq.push(std::make_pair(0, k));
        std::unordered_set<int> visited;
        int delay = 0;
        while (pq.size() > 0) {
            std::pair<int, int> current = pq.top();
            pq.pop();

            if (visited.contains(current.second)) {
                continue;
            }

            visited.insert(current.second);
            delay = std::max(delay, -current.first);

            if (visited.size() == n) {
                return delay;
            }

            for (auto [delay, destination] : adj[current.second]) {
                pq.push(std::make_pair(current.first + delay, destination));
            }
        }

        return -1;
    }
};
